sae/blocks.go

```diff
@@ -5,42 +5,241 @@ package sae
-	return nil, errUnimplemented
+// [VM.SetPreference] as the parent. The block context MAY be nil.
+func (vm *VM) BuildBlock(ctx context.Context, bCtx *block.Context) (*blocks.Block, error) {
+	parent := vm.preference.Load()
+
+	baseFee := uint256.NewInt(0)
+	if false { // Demonstrate the expected actual computation of `baseFee`
+		clock := parent.ExecutedByGasTime() // TODO(arr4n) carry the worst-case equivalent and use it instead
+		clock.BeforeBlock(vm.hooks, nil)    // TODO(arr4n) modify this to accept the FastForward() arguments, not the Header
+		baseFee = clock.BaseFee()
+	}
+
+	filter := txpool.PendingFilter{
+		BaseFee: baseFee,
+	}
+	return vm.buildBlock(ctx, bCtx, parent, vm.config.Now(), vm.mempool.TransactionsByPriority(filter))
+}
+
+var errExecutionLagging = errors.New("execution lagging for settlement")
+
+// buildBlock implements the block-building logic shared by [VM.BuildBlock] and
+// [VM.VerifyBlock].
+func (vm *VM) buildBlock(ctx context.Context, bCtx *block.Context, parent *blocks.Block, blockTime time.Time, candidates []*txgossip.LazyTransaction) (*blocks.Block, error) {
+	log := vm.log().With(
+		zap.Uint64("parent_height", parent.Height()),
+		zap.Stringer("parent_hash", parent.Hash()),
+		zap.Time("block_time", blockTime),
+	)
+
+	settleAt := unix(blockTime.Add(-params.Tau))
+	lastSettled, ok, err := blocks.LastToSettleAt(settleAt, parent)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		log.Warn("Execution lagging when determining last block to settle")
+		return nil, errExecutionLagging
+	}
+
+	// Although, at the time of writing this, the implementation of
+	// [blocks.LastToSettleAt] guarantees that the returned block has finished
+	// execution, this isn't a stated invariant and it's possible for it to be
+	// more aggressive in the future.
+	if !lastSettled.Executed() {
+		log.Warn(
+			"Execution lagging for settlement artefacts",
+			zap.Uint64("settling_height", lastSettled.Height()),
+		)
+		return nil, fmt.Errorf("%w: settling block %d: %T.WaitUntilExecuted(): %v", errExecutionLagging, lastSettled.Height(), lastSettled, err)
+	}
+
+	sdb, err := state.New(lastSettled.PostExecutionStateRoot(), vm.exec.StateCache(), nil)
+	if err != nil {
+		b := lastSettled
+		return nil, fmt.Errorf("state.New([post-execution state root %#x of block %d], ...): %v", b.PostExecutionStateRoot(), b.Height(), err)
+	}
+
+	// TODO(arr4n) worst-case tx validation will happen here. This is just a
+	// sketch of the process.
+	history := parent.WhenChildSettles(lastSettled)
+	var receipts types.Receipts
+	for _, b := range history {
+		receipts = append(receipts, b.Receipts()...)
+	}
+
+	clock := lastSettled.ExecutedByGasTime()
+	for _, b := range history {
+		clock.BeforeBlock(vm.hooks, b.Header())
+
+		var lim gas.Gas
+		for _, tx := range b.Transactions() {
+			lim += gas.Gas(tx.Gas())
+		}
+		if err := clock.AfterBlock(lim, vm.hooks, b.Header()); err != nil {
+			return nil, err
+		}
+	}
+
+	hdr := &types.Header{
```

Comment:
Probably not super actionable right now, but figured I'd comment it since it came up in my review.
- `Coinbase` on C-chain and subnet-evm is typically set to the Blackhole address `common.Address{1}`. changing that here would change how fee burning would need to be calculated externally, so we probably want to keep that the same.
- `Difficulty` on C-chain and subnet-evm we set this explicitly to `1`. I'm not sure that it really matters... But probably makes sense not to change that pattern when SAE activates.
- All the upgrade values should be populated (if the upgrade calls for them), `BlobGasUsed`, `ExcessBlobGas`, `ParentBeaconRoot`. `Withdrawals` doesn't matter because we override the serialization not to include them ever.